### PR TITLE
fix: ensure closed shards with pending data are still claimed

### LINKED ISSUE #643: aws_kinesis: DynamoDB entries for closed shards not cleaned up during pod termination
## Problem Description

When Kinesis scales down (merges shards) and Bento pods scale down simultaneously in Kubernetes, DynamoDB lease table entries for closed shards are not cleaned up. This causes false positive latency alerts as the monitoring detects "stale" entries with old timestamps but no active consumers.

## Root Cause

There is a race condition between shard cleanup and pod termination:

1. **Kinesis scales down** (e.g., 4 shards → 2 shards)
   - Shards are marked closed with `EndingSequenceNumber`

2. **Bento pods scale down** shortly after (e.g., 3 pods → 1 pod)
   - Pods receive SIGTERM before consumers finish processing closed shards
   
3. **Shutdown logic leaves orphaned entries**
   - Consumer enters `awsKinesisConsumerClosing` state
   - Calls `Checkpoint(final=true)` which removes `ClientID` and `LeaseTimeout` but keeps the row
   - DynamoDB entry remains with only `StreamID`, `ShardID`, `SequenceNumber`

4. **Rebalancing logic doesn't clean up finished shards**
   - `runBalancedShards()` excludes finished shards from rebalancing
   - No cleanup logic exists for finished shards

## Expected Behavior

When a shard is closed by Kinesis and fully consumed, its DynamoDB entry should be deleted, even if the pod is terminated before the consumer naturally completes.

## Current Workaround

Manual deletion of DynamoDB entries without ClientID for closed shards.

## Impact

- Recurring false positive latency alerts
- Manual intervention required after each Kinesis scaling event
- Has occurred multiple times in production (see references)

## Proposed Solution

1. **Fix shutdown logic**: When a pod shuts down, check if the shard iterator is empty (shard finished) and call `Delete()` instead of `Checkpoint(final=true)`

2. **Add periodic cleanup**: In the rebalancing loop, proactively delete DynamoDB entries for shards that Kinesis has marked as closed

## References

This issue has been reported in the Bento Slack community and has occurred in production environments multiple times requiring manual cleanup.

## Environment

- Bento version: Latest (main branch)
- Deployment: Kubernetes with HPA-based pod autoscaling
- Kinesis: Auto-scaling enabled
- DynamoDB: Used for checkpoint storage

### PR DESCRIPTION
**Pull request by matus-tomlein**
## Summary

Fixes #643 

Closed shards (with `EndingSequenceNumber`) were being skipped during rebalancing, even if they still had unprocessed records. This caused data loss when a consumer restarted before fully draining a closed shard.

### The Problem

When a shard is closed, the old logic would skip it entirely during rebalancing:

```go
for _, s := range allShards {
    if !isShardFinished(s) {  // ← Closed shards skipped!
        unclaimedShards[*s.ShardId] = ""
    }
}
```

This caused a race condition:

1. Consumer is reading from Shard-0
2. Kinesis scales down → Shard-0 gets CLOSED (EndingSequenceNumber set)
3. Consumer hasn't finished draining Shard-0 yet (still has records)
4. Rebalance runs:
   - isShardFinished(Shard-0) = true
   - Shard-0 is NOT added to unclaimedShards
   - No consumer gets spawned for it
5. Consumer pod restarts (or crashes)
6. Shard-0 is orphaned — nobody will ever finish consuming it

### The Fix

Now we also check if a closed shard has a checkpoint in DynamoDB. If a checkpoint exists, the shard hasn't been fully consumed yet (since checkpointer.Delete() removes it upon completion), so we include it in unclaimedShards:

```go
if !isShardFinished(s) || shardsWithCheckpoints[shardID] {
    unclaimedShards[shardID] = ""
}
```

### Test plan

Added unit tests for `isShardFinished()` covering open, closed, and edge cases

### CODE REVIEW

File: internal/impl/aws/input_kinesis.go [hunk 1]
```diff
@@ -686,6 +696,27 @@ func (k *kinesisReader) runBalancedShards() {
...
+				for _, claim := range claims {
+					if finishedShardIDs[claim.ShardID] {
+						if err := k.checkpointer.Delete(k.ctx, info.id, claim.ShardID); err != nil {
+							k.log.Errorf("Failed to clean up finished shard '%v' from client '%v': %v", claim.ShardID, clientID, err)
+						} else {
+							k.log.Debugf("Cleaned up finished shard '%v' previously claimed by client '%v'", claim.ShardID, clientID)
+						}
+					}
+				}
+			}
```
[the remark this answers is not in the document]
Comment (matus-tomlein) [reply internal/impl/aws/input_kinesis.go]: ✅ Fixed in commit f14390b.

The cleanup now checks for expired leases before deleting:

```go
isExpired := time.Since(claim.LeaseTimeout) > k.leasePeriod*2
if isExpired {
    k.checkpointer.Delete(...)
}
```

Additionally, when skipping cleanup due to an active lease, it now logs:
```go
k.log.Debugf("Skipping cleanup of finished shard '%v' from client '%v' (lease still active)", claim.ShardID, clientID)
```

This prevents interfering with active consumers while still cleaning up entries with expired leases.

File: internal/impl/aws/input_kinesis.go [hunk 2]
```diff
@@ -485,8 +485,18 @@ func (k *kinesisReader) runConsumer(wg *sync.WaitGroup, info streamInfo, shardID
...
 			case awsKinesisConsumerClosing:
 				reason = " because the pipeline is shutting down"
-				if _, err := k.checkpointer.Checkpoint(context.Background(), info.id, shardID, recordBatcher.GetSequence(), true); err != nil {
-					k.log.Errorf("Failed to store final checkpoint for stream '%v' shard '%v': %v", info.id, shardID, err)
+				// If the iterator is empty, the shard is finished and should be deleted
+				// rather than checkpointed. This prevents orphaned DynamoDB entries when
+				// pods are terminated after Kinesis closes shards but before consumption completes.
+				if iter == "" {
+					k.log.Debugf("Deleting checkpoint for finished shard during shutdown: stream '%v' shard '%v'", info.id, shardID)
+					if err := k.checkpointer.Delete(k.ctx, info.id, shardID); err != nil {
```
[the remark this answers is not in the document]
Comment (matus-tomlein) [reply internal/impl/aws/input_kinesis.go]: ✅ Fixed in commit 108a85f.

The Delete operation in the shutdown handler now uses `context.Background()`:

```go
if iter == "" {
    // Use context.Background() to ensure the delete completes even during shutdown
    if err := k.checkpointer.Delete(context.Background(), info.id, shardID); err != nil {
        k.log.Errorf("Failed to remove checkpoint for finished stream '%v' shard '%v': %v", info.id, shardID, err)
    }
}
```

This matches the pattern used for the Checkpoint call on line 503 and ensures cleanup completes even when `k.ctx` is cancelled during shutdown.

Note: The Delete in the periodic cleanup loop (line 739) correctly continues to use `k.ctx` since it runs during normal operation and should respect shutdown signals.

File: internal/impl/aws/input_kinesis.go [hunk 3]
```diff
@@ -686,6 +697,34 @@ func (k *kinesisReader) runBalancedShards() {
...
+			for _, s := range allShards {
+				if isShardFinished(s) {
+					finishedShardIDs[*s.ShardId] = true
+				}
+			}
+			for clientID, claims := range clientClaims {
+				for _, claim := range claims {
+					if finishedShardIDs[claim.ShardID] {
+						// Only clean up if lease has expired or there's no client (orphaned entry)
+						isOrphaned := clientID == ""
```
[the remark this answers is not in the document]
Comment (matus-tomlein) [reply internal/impl/aws/input_kinesis.go]: ✅ Excellent catch! Fixed in commit f14390b.

You're absolutely right - the `clientID == ""` check would never be true. I've implemented your suggested solution of adding a separate query for orphaned entries.

**New method added** (`input_kinesis_checkpointer.go`):
```go
func (k *awsKinesisCheckpointer) OrphanedShards(ctx context.Context, streamID string) ([]string, error) {
    scanRes, err := k.svc.Scan(ctx, &dynamodb.ScanInput{
        TableName:        aws.String(k.conf.Table),
        FilterExpression: aws.String("StreamID = :stream_id AND attribute_not_exists(ClientID)"),
        ...
    })
    ...
}
```

**Updated cleanup logic** now has two phases:
1. Clean up entries with expired leases (from `AllClaims()`)
2. Clean up orphaned entries without ClientID (from `OrphanedShards()`)

```go
// Phase 1: Expired leases
for clientID, claims := range clientClaims {
    if finishedShardIDs[claim.ShardID] && isExpired { ... }
}

// Phase 2: Orphaned entries
orphanedShards, err := k.checkpointer.OrphanedShards(k.ctx, info.id)
for _, shardID := range orphanedShards {
    if finishedShardIDs[shardID] { ... }
}
```

This properly handles entries created by `Checkpoint(final=true)` during pod termination.

File: internal/impl/aws/input_kinesis.go [hunk 4]
```diff
@@ -485,8 +485,19 @@ func (k *kinesisReader) runConsumer(wg *sync.WaitGroup, info streamInfo, shardID
...
+					k.log.Debugf("Deleting checkpoint for finished shard during shutdown: stream '%v' shard '%v'", info.id, shardID)
+					// Use context.Background() to ensure the delete completes even during shutdown
+					if err := k.checkpointer.Delete(context.Background(), info.id, shardID); err != nil {
+						k.log.Errorf("Failed to remove checkpoint for finished stream '%v' shard '%v': %v", info.id, shardID, err)
+					}
+				} else {
+					if _, err := k.checkpointer.Checkpoint(context.Background(), info.id, shardID, recordBatcher.GetSequence(), true); err != nil {
+						k.log.Errorf("Failed to store final checkpoint for stream '%v' shard '%v': %v", info.id, shardID, err)
+					}
 				}
```
[the remark this answers is not in the document]
Comment (matus-tomlein) [reply internal/impl/aws/input_kinesis.go:490]: ✅ Documented in commit f14390b.

Great point about this edge case. I've added explicit documentation explaining this scenario:

```go
case awsKinesisConsumerClosing:
    // If the iterator is empty, the shard is finished and should be deleted
    // rather than checkpointed. This prevents orphaned DynamoDB entries when
    // pods are terminated after Kinesis closes shards but before consumption completes.
    //
    // Note: There's an edge case where if a shard naturally finishes (state becomes
    // awsKinesisConsumerFinished) but the pod is terminated before this defer runs,
    // the state will be changed to awsKinesisConsumerClosing. This means Delete could
    // be called from both code paths, but this is safe because Delete is idempotent.
    if iter == "" {
        k.checkpointer.Delete(context.Background(), ...)
    }
```

The comment now clearly documents that:
1. This double-deletion scenario can occur
2. It's safe due to DynamoDB DeleteItem idempotency
3. The timing window where this happens (between state change and defer execution)

I chose documentation over refactoring the state logic since:
- The current behavior is correct and safe
- Preserving state across context cancellation would add complexity
- The idempotent Delete makes this a non-issue functionally

File: internal/impl/aws/input_kinesis.go [hunk 5]
```diff
@@ -686,6 +697,34 @@ func (k *kinesisReader) runBalancedShards() {
...
+						if isOrphaned || isExpired {
+							if err := k.checkpointer.Delete(k.ctx, info.id, claim.ShardID); err != nil {
+								k.log.Errorf("Failed to clean up finished shard '%v' from client '%v': %v", claim.ShardID, clientID, err)
+							} else {
+								k.log.Debugf("Cleaned up finished shard '%v' previously claimed by client '%v'", claim.ShardID, clientID)
+							}
+						}
+					}
+				}
+			}
```
[the remark this answers is not in the document]
Comment (matus-tomlein) [reply internal/impl/aws/input_kinesis.go]: ✅ Fixed in commit f14390b.

You're absolutely right about the potential for log confusion. I've added detailed debug logging to make the behavior clear:

**When skipping cleanup due to active lease:**
```go
k.log.Debugf("Skipping cleanup of finished shard '%v' from client '%v' (lease still active)", claim.ShardID, clientID)
```

**When cleaning up expired lease:**
```go
k.log.Debugf("Cleaned up finished shard '%v' with expired lease from client '%v'", claim.ShardID, clientID)
```

**When cleaning up orphaned entry:**
```go
k.log.Debugf("Cleaned up orphaned finished shard '%v'", shardID)
```

This makes it easy to understand from logs:
1. Why cleanup was skipped (active lease protection)
2. Which cleanup phase succeeded (expired lease vs orphaned)
3. The timeline of events during a rebalance

The race you described (consumer completes → Checkpoint(final=true) → next rebalance deletes) will now be visible in logs, and the duplicate Delete attempts are safe due to idempotency.

**Comment by matus-tomlein** [comment]
@jem-davies @gregfurman We have simplified the PR and addressed all the comments above – could you please give this another look?

**Comment by jem-davies** [comment]
Closed as included in #680
